src/org/jetbrains/plugins/scala/codeInspection/monads/StatefulMonads.scala

```diff
@@ -0,0 +1,27 @@
+package org.jetbrains.plugins.scala.codeInspection.monads
+
+import com.intellij.psi.PsiClass
+import org.jetbrains.plugins.scala.lang.psi.types.{ScDesignatorType, ScType}
+
+import scala.concurrent.Future
+import scala.util.Try
+
+/**
+ * @author Sergey Tolmachev (tolsi.ru@gmail.com)
+ * @since 29.09.15
+ */
+object StatefulMonads {
+  private[monads] final val RootTypeCanonicalNamePrefix = "_root_."
+  private[monads] final lazy val StatefulMonadsTypes: Set[Class[_]] = Set(classOf[Future[_]], classOf[Try[_]])
+  private[monads] final lazy val StatefulMonadsTypesNames = StatefulMonadsTypes.map(t => RootTypeCanonicalNamePrefix + t.getCanonicalName)
+
+  // todo I hate myself :<
+  private[monads] def isStatefulMonadType(t: ScType): Boolean = {
+    val psiEl = t.unpackedType.asInstanceOf[ScDesignatorType].element.asInstanceOf[PsiClass]
+    val superTypes = psiEl.getImplementsListTypes
+    val selfTypeAndSuperTypesNames = superTypes.map(_.getCanonicalText).map(t => RootTypeCanonicalNamePrefix + cleanTypeParams(t)) :+ cleanTypeParams(t.canonicalText)
+    selfTypeAndSuperTypesNames.exists(StatefulMonadsTypesNames.contains)
```

Comment:
@niktrop how can I check that expression if of type Future of Try[Any] with this method, not simply Future? If I can't, then how can I use it with ScType? This method is similar to my hack.